From 66e0954703ac47028e31e13cfc4e08d93f1e7638 Mon Sep 17 00:00:00 2001
From: Noam Lessner <nlessner@habana.ai>
Date: Wed, 05 Feb 2020 09:22:31 +0200
Subject: [PATCH] [SW-8229] verify fail when no valid key found

return verification fail if no key found that fits
non-revoked hashes in fues. According to specs only first
customer key and habana key are validated.

Change-Id: Id76a14cad771520ec912761d72ea3ae1dbbb7065
Signed-off-by: Noam Lessner <nlessner@habana.ai>
---

diff --git a/common/secureboot.c b/common/secureboot.c
index 52b39b1..ca8c9a0 100644
--- a/common/secureboot.c
+++ b/common/secureboot.c
@@ -32,6 +32,7 @@
 {
 	int verify = 0;
 	int err = 0;
+	int key_scan;
 	unsigned char block_buf[SECUREBOOT_HASH_BLOCK_CALC_BUF];
 	unsigned char test_phase0a_hash[SECURITY_KEY_HASH_ARRAY_BIT_SIZE];
 	uint16_t test_phase0b_data_len = image_size;
@@ -50,8 +51,8 @@
 	    SECURITY_FW_VERSION_SECURE)
 		return SECUREBOOT_VERIFY_FAILED;
 
-	for (int key_scan = SECUREBOOT_CUSTOMER_KEY;
-	     key_scan >= SECUREBOOT_HABANA_KEY; key_scan--) {
+	for (key_scan = SECUREBOOT_CUSTOMER_KEY; key_scan >=
+	     SECUREBOOT_HABANA_KEY; key_scan--) {
 		if (security_get_key_hash((uint32_t *)test_phase0a_hash,
 					  key_scan)) {
 			/* check pub key hash versus efuse */
@@ -61,22 +62,15 @@
 						     test_phase0a_hash,
 						     &verify, block_buf,
 						     sizeof(block_buf));
-			if (!verify && (key_scan == SECUREBOOT_HABANA_KEY)) {
-				regmap_writel(err,
-					      mmPSOC_SECURITY_SW_STATUS_0);
-				regmap_writel(verify,
-					      mmPSOC_SECURITY_SW_STATUS_1);
-				return SECUREBOOT_VERIFY_FAILED;
-			}
-		} else {
-			/* if (key_scan == SECUREBOOT_HABANA_KEY) &&
-			 *(security_all_customer_revoked()){
-			 *	security_check_force_habana_key();
-			 *}
-			 */
 		}
 	}
-	/* TODO: add if for loop doesn't find a valid key hash in fuse */
+	/* if for loop doesn't find a valid key hash in fuse, return error */
+	if (!verify && (key_scan == SECUREBOOT_HABANA_KEY)) {
+		regmap_writel(err, mmPSOC_SECURITY_SW_STATUS_0);
+		regmap_writel(verify, mmPSOC_SECURITY_SW_STATUS_1);
+		return SECUREBOOT_VERIFY_FAILED;
+	}
+
 
 	/* Phase0b uses the public key (test_public_key) to validate signature
 	 * (test_phase0b_sig) of the data blob (test_phase0b_data)
